src/main/java/com/minecolonies/coremod/colony/crafting/CustomRecipeFactory.java

```diff
@@ -0,0 +1,274 @@
+    }
+
+    @NotNull
+    @Override
+    public TypeToken<CustomRecipe> getFactoryOutputType()
+    {
+        return TypeToken.of(CustomRecipe.class);
+    }
+
+    @NotNull
+    @Override
+    public TypeToken<FactoryVoidInput> getFactoryInputType()
+    {
+        return TypeConstants.FACTORYVOIDINPUT;
+    }
+
+    private CompoundNBT getCompoundForItemStack(final ItemStack is)
+    {
+        final CompoundNBT item = new CompoundNBT();
+        item.putString(ITEM_PROP, is.getItem().getRegistryName().toString());
+        item.putInt(COUNT_PROP, is.getCount());
+        if(is.hasTag())
+        {
+            item.put(RECIPE_TAG, is.serializeNBT());
+        }
+        return item;
+    }
+
+    private ItemStack getItemStackForCompound(final CompoundNBT nbt)
+    {
+        final ItemStack item = new ItemStack(ForgeRegistries.ITEMS.getValue(new ResourceLocation(nbt.getString(ITEM_PROP))), nbt.getInt(COUNT_PROP));
+        if(nbt.hasUniqueId(RECIPE_TAG))
+        {
+            item.deserializeNBT(nbt.getCompound(RECIPE_TAG));
+        }
+        return item;
+    }
+
+    @NotNull
+    @Override
+    public CompoundNBT serialize(@NotNull final IFactoryController controller, @NotNull final CustomRecipe recipe)
+    {
+        // CustomRecipes involve a large number of ItemStacks, and a number of data types.  Individually, this isn't that big of a deal (~1.2KB vs 120bytes).
+        // However, for large sets, such as CustomRecipeManager, this can grow into a very large total (ie, default Minecolonies + Structurize recipes alone can total >600KB).
+        // If transmitting a large number of recipes at once across a network, favor the byte-based serializer instead.
+        // This serialization also populates the RecipeStorage inside the Custom Recipe, which remains cached until the next data pack reload.
+        // Be aware of the 2MB limit for CompoundNBTs if assigning a large number of these together into a listNBT.
+        final CompoundNBT compound = new CompoundNBT();
+        compound.putString(RECIPE_CRAFTER_PROP, recipe.getCrafter());
+        compound.putString(CUSTOM_RECIPE_ID_PROP, recipe.getRecipeStorage().getRecipeSource().toString());
+        if(recipe.getRequiredResearchId() != null)
+        {
+            compound.putString(RECIPE_RESEARCHID_PROP, recipe.getRequiredResearchId().toString());
+        }
+        if(recipe.getExcludedResearchId() != null)
+        {
+            compound.putString(RECIPE_EXCLUDED_RESEARCHID_PROP, recipe.getExcludedResearchId().toString());
+        }
+        compound.putInt(RECIPE_BUILDING_MIN_LEVEL_PROP, recipe.getMinBuildingLevel());
+        compound.putInt(RECIPE_BUILDING_MAX_LEVEL_PROP, recipe.getMaxBuildingLevel());
+        compound.putBoolean(RECIPE_MUST_EXIST, recipe.getMustExist());
+        compound.putBoolean(RECIPE_SHOW_TOOLTIP, recipe.getShowTooltip());
+        compound.put(RECIPE_TYPE_RECIPE, controller.serialize(recipe.getRecipeStorage()));
```

Comment:
Why do we need to serialize this at all?

Reply:
I'd actually go so far as to say you SHOULDN'T serialize this. It should be built/cached when somebody calls getRecipeStorage off the deserialized customRecipe downstream.

Reply:
Ok, I found why. You're not serializing the core collections, you're depending on recipeStorage to do that. I think that's a bad idea, because it creates the possibility of a de-synced central ID by serializing a cached item which contains an ID that is not included in hashcode/equals.

Reply:
I don't know that we do; my original thought process was to serialize the individual components, leaving the RecipeStorages alone; see discussion here (https://github.com/ldtteam/minecolonies/pull/6943#discussion_r615279963). I'm not sure what the consequences of even merely calling `getRecipeStorage` are, and possible desync is a situation (though wasn't sure about the consequences of throwing away old itokens during a data pack reload on a server).

The byte-focused serialization/deserialization operates in that way, and is currently the actually called code; I can switch the nbt-variant back.

Reply:
Yeah, make the nbt-variant match the byte-focused, and please remove the constructor that takes a recipestorage.